### PR TITLE
Muhammad/Update Colonist Dev Repo

### PR DESCRIPTION
**Pull request by a-knaw-knee-mus**
## Summary
- Update Repo to create a landing page for future candidates during their take home test

## References
- closes https://github.com/colonistio/katan/issues/16900

## Checklists

- [x] This PR abides by [Dev-lvl-1](https://github.com/colonistio/katan/blob/development/docs/process/developer-level-1.md), [Dev-lvl-2](https://github.com/colonistio/katan/blob/development/docs/process/developer-level-2.md) & [Dev-lvl-3](https://github.com/colonistio/katan/blob/development/docs/process/developer-level-3.md) & [Coding Rules](https://github.com/colonistio/katan/blob/development/docs/process/coding.md)
- [x] I reviewed my own Pull Request commit by commit
- [x] I didn't just select everything, this PR really does abide by these ^

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,8 @@
-# developer-checklist
\ No newline at end of file
+# Developer Checklist
+
+## Open the Perfect PR
+1. Create a new branch `your_short_name/feature-description` (Example: `demi/fixing-duplicate-port-bug`). When making a branch, we use our names to distinguish who does the feature.
+1. The PR itself should also begin with `your_short_name/<PR title>` for similar reasons
+    - PR Title should be as short and concise as possible
+1. PRs submitted during your time at Colonist follow a strict format and criteria, both for the [PR description](/docs/pr-format.md), and code quality
+1. Before submitting a PR, ensure it abides by [Developer fundamentals](/docs/developer/developer-fundamentals.md), [Developer Intermediate](/docs/developer/developer-intermediate.md), and [Developer advanced](/docs/developer/developer-advanced.md)
```
Comment (demiculus) [inline README.md]: 1. We need to give them instructions on what steps to follow.
2. We don't want the instructions on readme since we want this repo to be a more generic repo for other teams & companies to reffer.

Next steps:
1. Edit this read me explaining what the repo is
2. Create a new doc named `Colonist Take Home Test Opening PR` and add the stuff there.

Comment (demiculus) [reply README.md]: `Colonist Take Home Test Opening PR` should be more clear.

1. Create an empty repository
2. Create a branch
3. Add the code that you wrote in JSFiddle to that branch
4. Push your code to the repo
5. Create a PR that merges to master
6. Make sure repo is public
7. Make sure everything follows these guides
8. Send the PR link to your contact for review

Comment (demiculus) [inline README.md]: developer/developer-...

The repo is also named developer. Remove the unnecessary developers.

File: docs/developer/developer-fundamentals.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
+# Developer Fundamentals
```
Comment (demiculus) [inline docs/developer/developer-fundamentals.md]:
```suggestion
# Fundamentals
```

Comment (demiculus) [reply docs/developer/developer-fundamentals.md]: remove the othe unnecessarry developers

File: docs/developer/developer-fundamentals.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+## Table of contents
+- [Developer Level-1](#developer-fundamentals)
```
Comment (demiculus) [inline docs/developer/developer-fundamentals.md]: still says level-1

File: docs/developer/developer-fundamentals.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  - [Don't Use Magic Numbers](#dont-use-magic-numbers)
+  - [Avoid Code Repetition](#avoid-code-repetition)
+  - [Avoid Tight Coupling](#avoid-tight-coupling)
+  - [Iterate Before Opening a PR](#iterate-before-opening-a-pr)
+  - [Resolving Conversations](#resolving-conversations)
+  - [When to Merge PR](#when-to-merge-pr)
+
+## Importance of reading the material
+
+-  Most people who skim through the reading materials tend to fail the trial. The key to passing is going through both [Developer Fundamentals](#developer-fundamentals) & [Developer Intermediate](./developer-intermediate.md) during each PR for the first 2 weeks.
```
Comment (demiculus) [inline docs/developer/developer-fundamentals.md]:
```suggestion
-  Most people who skim through the reading materials tend to fail at becoming long term contributors. The key to passing is going through the developer docs during each PR for the first 2 weeks. And coming back to them once a month to realign.
```

Comment (demiculus) [reply docs/developer/developer-fundamentals.md]: Change it to something like this. 
There is a forgetting curve. People need to come back in certain intervals.

File: docs/developer/developer-fundamentals.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+## Naming Conventions
+
+### Branch Naming
+
+- `your_short_name/feature-description`
+- Ex: `goktug/main-ejs-env-system`
+- Follow https://mirrors.edge.kernel.org/pub/software/scm/git/docs/git-check-ref-format.html, and don't use any special characters like `'`, just to be safe.
```
Comment (demiculus) [inline docs/developer/developer-fundamentals.md]:
```suggestion
- Follow https://mirrors.edge.kernel.org/pub/software/scm/git/docs/git-check-ref-format.html, and don't use any special characters like `'`.
```

File: docs/pr-format.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  - [ ] Mobile
+
+## References
+<!-- LINK EVERY SINGLE RELATED ISSUE OR PR -->
+* closes #
+* [example_url](www.google.com)
+
+## Checklists
+<!-- DO NOT DELETE OR EDIT THIS LIST -->
+- [ ] This PR abides by [Dev-lvl-1](../tree/development/docs/process/developer-level-1.md), [Dev-lvl-2](../tree/development/docs/process/developer-level-2.md) & [Dev-lvl-3](../tree/development/docs/process/developer-level-3.md) & [Coding Rules](../tree/development/docs/process/coding.md)
```
Comment (demiculus) [inline docs/pr-format.md]: edit this accordingly?

**Review by demiculus — CHANGES_REQUESTED** [review]
Will you do the intermediate later?

**Comment by a-knaw-knee-mus** [comment]
Both intermediate and advanced yes

**Comment by a-knaw-knee-mus** [comment]
@demiculus  Should all references to patch notes, and adding to patch notes be removed?

**Comment by demiculus** [comment]
> @demiculus Should all references to patch notes, and adding to patch notes be removed?

I'm not sure. My thought pattern:

- This repository is a copy of Colonist
- Ideally we have single place that has all the mental models rather than 2 places
- It might make more sense having the mental models here and deleting the ones in Colonist
- Colonist specific & generic stuff should be separated

Overall I see 2 options
1. Move the Colonist specific stuff to a different doc in this repo
2. Have a separate doc in Colonist repo for Colonist specific stuff

I think 2nd option makes more sense. 
But I want to get Gokus opinion about this before moving forward with editing Colonist repo.

Tldr: Remove Colonist specific stuff from this repo

File: docs/pr-quality/pr-format.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+## Details
+* edit_me_pls
+
+## Testing
+* explain_how_qa_should_test_this_PR
+* list_edge_cases_qa_should_extra_pay_attention_to:
+- [ ] edge_case_1
+- [ ] edge_case_2
+#### Platforms affected by this PR:
```
Comment (JuanRoman77) [inline docs/pr-quality/pr-format.md:17]:
```suggestion
#### Platforms affected by this PR:  
<!-- DO NOT DELETE OR EDIT THIS LIST -->
```

File: docs/code-quality/advanced.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	- [Don't Overengineer](#dont-overengineer)
+	- [Use Existing Code](#use-existing-code)
+	- [Understand Impact](#understand-impact)
+	- [Suggesting A Library](#suggesting-a-library)
+	- [Knowing What You Don't Know](#knowing-what-you-dont-know)
+	- [Aware of Impact](#aware-of-impact)
+
+## Large Update Rules
+
+Any issue that requires more than 10 commits or 50 lines of code can be considered a large update (unless it is pure refactor)
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: - `a pure refactor`
- period at the end

File: docs/code-quality/advanced.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	- [Write Tests](#write-tests)
+	- [Force Push](#force-push)
+	- [Don't Overengineer](#dont-overengineer)
+	- [Use Existing Code](#use-existing-code)
+	- [Understand Impact](#understand-impact)
+	- [Suggesting A Library](#suggesting-a-library)
+	- [Knowing What You Don't Know](#knowing-what-you-dont-know)
+	- [Aware of Impact](#aware-of-impact)
+
+## Large Update Rules
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: Workflow > Rules?

File: docs/code-quality/advanced.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+Any issue that requires more than 10 commits or 50 lines of code can be considered a large update (unless it is pure refactor)
+
+1. If you're doing a big feature, first create a branch and do a rough version of the feature that works, without worrying much about commit readability and architecture
+2. Push that branch to Github as a draft PR
+3. Review your commit at PR UI from the eye of reviewer
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: Review your PR on Github, as if you were the actual reviewer

File: docs/code-quality/advanced.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	- [Understand Impact](#understand-impact)
+	- [Suggesting A Library](#suggesting-a-library)
+	- [Knowing What You Don't Know](#knowing-what-you-dont-know)
+	- [Aware of Impact](#aware-of-impact)
+
+## Large Update Rules
+
+Any issue that requires more than 10 commits or 50 lines of code can be considered a large update (unless it is pure refactor)
+
+1. If you're doing a big feature, first create a branch and do a rough version of the feature that works, without worrying much about commit readability and architecture
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: - Can remove `If you're doing a big feature`, since we assume you're doing a big feature if you're following this list
- Can remove `first`, since this is already the first item
- Specify local branch

So:

`1. Create a local branch and do a rough version of the feature that works, without worrying much about commit readability and architecture`

File: docs/code-quality/advanced.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+2. Push that branch to Github as a draft PR
+3. Review your commit at PR UI from the eye of reviewer 
+4. Give yourself stuff to fix, like change var name, separate variables, split function into multiple, do better architecture, etc..
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: Find things to improve during your review, such as: changing variable names, separating variables, splitting functions, better architecture, etc.

File: docs/code-quality/advanced.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+5. Then create a todo list, and how could the architecture have been done cleanly
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: Confused by this

File: docs/code-quality/advanced.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+6. Create a new branch
+7. Start moving code to new branch and rewrite everything but in a clean, final way
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: to _the_ new branch, and rewrite everything cleanly, to prepare for your final PR submission

File: docs/code-quality/advanced.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+8. If this feature will require multiple PRs, then create a main PR, then write code in parts in the PRs.
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: then create one main PR, and create a sub-PR for each self-contained section of the feature. Then merge each sub-PR into the main PR, which will ultimately be reviewed and merged

File: docs/code-quality/advanced.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+	- Example: https://prnt.sc/2011xo4
+	- Video explanation: https://www.loom.com/share/703185b81607482ea304838064d34578
+9. If a code/logic is hard to understand first create a PR that refactors that part of the code, once its approved create the new feature
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: If _any_ code/logic is hard to understand, first create a PR to refactors that part of the code. Once _it's_ approved, you can move on to create the new feature

File: docs/code-quality/advanced.md [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+10. If you have to merge development branch please make sure to follow following. It will help to track commits properly
+    - squash merge them
+	- merge it in the main branch only
+	- if you need development in sub branch, merge it in the main branch first and then create a sub branch from it
+	- Bad example: https://prnt.sc/9gbezql7iQ0P
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: This section is confusing

File: docs/code-quality/advanced.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+## Add Comments
+
+We try to avoid comments as much as possible but under certain circumstances they need to be added.
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: , but

File: docs/code-quality/advanced.md [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+## Write Tests
+
+Complex logic should always be tested 
+Examples: https://prnt.sc/20127k6 https://prnt.sc/20127s7
+
+## Force Push
+
+- Do not force push on branches that have received feedback
+- Instead fork your previous branch, create a new feedback branch and do the updates there and merge into your original branch. https://www.loom.com/share/9bf9e79e21bc442498182571c6d62517
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: Instead, create a feedback branch from your main branch, and commit your updates there. Then merge the feedback branch into your main branch.

File: docs/code-quality/advanced.md [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+
+## Don't Overengineer
+
+We're trying to solve problems we have right now. We don't want to solve theoretical problems or solve wide variety of problems that open source libraries have to do.
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: - solve a wide variety
- open source libraries already do

File: docs/code-quality/advanced.md [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+## Force Push
+
+- Do not force push on branches that have received feedback
+- Instead fork your previous branch, create a new feedback branch and do the updates there and merge into your original branch. https://www.loom.com/share/9bf9e79e21bc442498182571c6d62517
+
+## Don't Overengineer
+
+We're trying to solve problems we have right now. We don't want to solve theoretical problems or solve wide variety of problems that open source libraries have to do. 
+
+Example conversation of overengineering: https://prnt.sc/o2fSYom1C3tq
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md:57]: new line after this

File: docs/code-quality/advanced.md [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+## Use Existing Code
+
+When you're writing a new code, ask yourself if there could be anything like this in the codebase. Search for it and use that one. We don't want duplicate logic.
```
Comment (rgsthethird) [inline docs/code-quality/advanced.md]: When you're writing new code, ask yourself if there could be anything like it in the codebase. Search for it and use the existing code. We don't want to duplicate logic.

File: docs/code-quality/coding-rules.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
+# Coding
+## General Rules
+
+1. Never work on the master or staging branches
+2. Always work on a custom branch dedicated to your feature
+5. Try to play around the code and discover how it works by yourself as much as possible
+7. If you just remembered a line of code that should've been added to the last commit, you should amend that line to last commit here: https://prnt.sc/v0t4ld
```
Comment (rgsthethird) [inline docs/code-quality/coding-rules.md]: numbers

Comment (rgsthethird) [inline docs/code-quality/coding-rules.md]: to the last commit

File: docs/code-quality/coding-rules.md [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+
+## Code Complete Tips
```
Comment (rgsthethird) [inline docs/code-quality/coding-rules.md]: I don't understand this section title

File: docs/code-quality/fundamentals.md [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+  - [Don't Use Magic Numbers](#dont-use-magic-numbers)
+  - [Avoid Code Repetition](#avoid-code-repetition)
+  - [Avoid Tight Coupling](#avoid-tight-coupling)
+  - [Iterate Before Opening a PR](#iterate-before-opening-a-pr)
+  - [Resolving Conversations](#resolving-conversations)
+  - [When to Merge PR](#when-to-merge-pr)
+
+## Importance of reading the material
+
+-  Most people who skim through the reading materials tend to fail at becoming long term contributors. The key to passing is going through the developer docs during each PR for the first 2 weeks. And coming back to them once a month to realign.
```
Comment (rgsthethird) [inline docs/code-quality/fundamentals.md]: Most people who skim through the reading materials fail to become long term contributors. The key to passing is reading the developer docs during each PR for the first 2 weeks, until you've internalized them. And read them again each month to realign.

File: docs/code-quality/intermediate.md [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+- Desktop: 1440x900
+
+even if [1920x1080 is the most widely desktop screen today](https://www.altamira.ai/blog/common-screen-sizes-for-responsive-web-design/)we should still be designing for 1440x900. Reason for this is that some players are playing in a smaller browser screen compared to their desktop and its a good practice to [design in a small screen and make it bigger](https://thewhitelabelagency.com/recommended-screen-resolution-for-web-design/) compared to designing it big and making it responsive to a smaller screen
+
+Consider different screen sizes: https://prnt.sc/w0vSLSNHRTLk
+
+### Unique Screen Formats
+- Discord Game Labs (https://prnt.sc/RqNuctEblZ0A)doesn't allow full screen so we shouldn't do our usual format of 1440x900. Its better to design in a smaller format of 1280x720.
+
+We might design some other screen with unique formats in the future. Keep in mind that it should be responsive and workable on that format
```
Comment (rgsthethird) [inline docs/code-quality/intermediate.md]: I think we can remove this

File: docs/pr-quality/colonist-take-home-test-opening-pr.md [hunk 27]
```diff
@@ -0,0 +1,12 @@
...
+1. Create an empty Github repository
+1. Pull repo
+1. Checkout a new branch following proper [naming conventions](/docs/code-quality/fundamentals.md#branch-naming)
+1. Add the code that you wrote in JSFiddle to that branch. Ensure the JSFiddle link is public
+1. Push branch
+    - Ensure code abides by the developer [fundamentals](/docs/code-quality/fundamentals.md), [intermediate](/docs/code-quality/intermediate.md), and [advanced](/docs/code-quality/advanced.md) practices
+1. Create a PR that merges to master following proper [naming conventions](/docs/code-quality/fundamentals.md#pr-naming)
+    - PRs must follow a strict criteria; refer to [PR format](/docs/pr-quality/pr-format.md)
+1. Ensure repo is public
+1. Send the PR link to your contact for review
```
Comment (rgsthethird) [inline docs/pr-quality/colonist-take-home-test-opening-pr.md]: Numbers

File: docs/pr-quality/colonist-take-home-test-opening-pr.md [hunk 28]
```diff
@@ -0,0 +1,12 @@
+# Colonist Take Home Test Opening PR
+
+1. Create an empty Github repository
+1. Pull repo
```
Comment (rgsthethird) [inline docs/pr-quality/colonist-take-home-test-opening-pr.md]: _Fork this_ repo?

File: docs/pr-quality/colonist-take-home-test-opening-pr.md [hunk 29]
```diff
@@ -0,0 +1,12 @@
...
+1. Checkout a new branch following proper [naming conventions](/docs/code-quality/fundamentals.md#branch-naming)
+1. Add the code that you wrote in JSFiddle to that branch. Ensure the JSFiddle link is public
+1. Push branch
```
Comment (rgsthethird) [inline docs/pr-quality/colonist-take-home-test-opening-pr.md]: Push your local branch to your new repo

File: docs/pr-quality/colonist-take-home-test-opening-pr.md (hunk 27 above)
Comment (a-knaw-knee-mus) [reply docs/pr-quality/colonist-take-home-test-opening-pr.md]: The list will automatically number them as expected if reading in 'preview' format. Should it still use consecutive numbering?

1. First
1. Second

Comment (rgsthethird) [reply docs/pr-quality/colonist-take-home-test-opening-pr.md]: yeah that would be better, unless that messes up the formatting in the preview format

Comment (a-knaw-knee-mus) [reply docs/pr-quality/colonist-take-home-test-opening-pr.md]: Sure will change to consecutive numbers. It doesn't mess up formatting, this is just another way of making consecutive lists I use.

**Review by rgsthethird — APPROVED** [review]
